schema/chat_room.proto

```diff
@@ -0,0 +1,343 @@
+  ];
+}
+
+message UpdateRoomRequest {
+  string name = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+
+  Room room = 2;
+}
+
+message DeleteRoomRequest {
+  string name = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+}
+
+message ListRoomsRequest {
+  int32 page_size = 1;
+
+  string page_token = 2;
+}
+
+message ListRoomsResponse {
+  repeated Room rooms = 1;
+
+  string next_page_token = 2;
+}
+
+message StreamRoomRequest {
+  string name = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+}
+
+message SearchAllRequest {
+  string query = 1 [(google.api.required) = true];
+}
+
+message SearchAllMetadata {
+  google.rpc.RetryInfo retry_info = 1;
+}
+
+message SearchAllResponse {
+  repeated Message messages = 1;
+}
+
+message Message {
+  string parent = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room"];
+
+  string name = 2 [(google.api.resource).path = "messages/*"];
+
+  // This will eventually become a User message that will cover add coverage for
+  // shared resources.
+  string author = 3;
+
+  string content = 4;
+
+  bytes image = 5;
+
+  google.protobuf.Timestamp create_time = 6;
+
+  google.protobuf.Timestamp update_time = 7;
+}
+
+message CreateMessageRequest {
+  string parent = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+
+  Message message = 2;
+}
+
+message GetMessageRequest {
+  string parent = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+
+  string name = 2 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Message",
+    (google.api.required) = true
+  ];
+}
+
+message UpdateMessageRequest {
+  string parent = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+
+  string name = 2 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Message",
+    (google.api.required) = true
+  ];
+
+  Message message = 3;
+}
+
+message DeleteMessageRequest {
+  string parent = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+
+  string name = 2 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Message",
+    (google.api.required) = true
+  ];
+}
+
+message ListMessagesRequest {
+  string parent = 1 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Room",
+    (google.api.required) = true
+  ];
+
+  string name = 2 [
+    (google.api.resource_type) = "google.showcase.v1alpha3.Message",
+    (google.api.required) = true
+  ];
+
+  int32 page_size = 3;
+
+  string page_token = 4;
+}
+
+message ListMessagesResponse {
+  repeated Message messages = 1;
+
+  string next_page_token = 2;
+}
+
+message ChatroomBehaviorHeader {
```

Comment:
What is this thing?

Reply:
This is an idea that @pongad tossed around in one of our testing meetings. This message will be serialized and passed in as metadata alongside any of the requests. If `DEFAULT` or `BEHAVIOR_UNSPECIFIED` or not passed, the showcase server will work as expected. The server will keep track of all messages, rooms and messages created.

If `STATELESS` is passed, the server will have specific expected request values and responses. RPCs will work with correct values but will not have the side-effect of changing state. ie The Server will always have the same rooms and messages. This will make it such that a test that uses a specific method, will not need any setup code. and the tests written on top of it will not have to worry about parallel tests causing an invalid state in the test environment. 

After writing this paragraph, I am on the 80% sure that this is unnecessary to be honest. I am not really thinking that there is gain to having a stateless solution.

Reply:
Yeah, this sounds like YAGNI to me (and confusing YAGNI at that).
I _think_ what you are trying to get here is the ability to not have to do "build up" in order to run specific tests? I assert you could get 90% of that effect by just having a documented shortcut, e.g. "If you start Showcase with the `--create-scaffolding` switch, it automatically creates three rooms for you."

(And you might not even care about _that._)